docs/user_guide/ml.md

```diff
@@ -1,8 +1,254 @@
+It is of good practice to use `task.target` to be that your "machine-learning-task" related files are saved in the same directory.
+In this example we want to save each fold separately e.g:
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def output
+:end-at: return target
+```
+
+
+## open_model:
+In the {py:meth}`~columnflow.ml.MLModel.open_model` function, you implement the loading of the trained model and if necessary its configuration, so it is ready to use for {py:meth}`~columnflow.ml.MLModel.evaluate`.
+This does not define how the model is build for {py:meth}`~columnflow.ml.MLModel.train`.
+
+The `target` parameter represents the local path to the models directory.
+In the the following example a TensorFlow model saved with Keras API is loaded and returned, and no further configuration happens:
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def open_model
+:end-at: return loaded_model
+```
+
+## train:
+In the {py:meth}`~columnflow.ml.MLModel.train` function, you implement the initialization of your models and the training loop.
+The `task` corresponding to the models training is {py:class}`~columnflow.tasks.ml.MLTraining`.
+By default {py:meth}`~columnflow.ml.MLModel.train` has access to the location of the models `inputs` and `outputs`.
+
+In columnflow, k-fold is enabled by default.
+The `Self` argument in {py:meth}`~columnflow.ml.MLModel.train` referes to the instance of the fold.
+Using `Self`, you have also access to the entire `analysis_inst`ance the `config_inst`ance of the current fold, and to all the derived parameters of your model.
+
+With this information, you can call and prepare the columns to be used by the model for training.
+In the following exmaple a very simple dummy trainings loop is performed using Keras fit function.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def train
+:end-at: return
+```
+### Good practice for training:
+It is considered to be a good practice to define helper functions for model building and column preprocessing and call them in {py:meth}`~columnflow.ml.MLModel.train`.
+The reasoning behind this design decision is that in {py:meth}`~columnflow.ml.MLModel.train`, you want to focus more on the definition of the actual trainings loop.
+Another reason is that especially the preparation of events, can take a large amount of code lines, making it messy to debug.
+In the following example it is shown how to define these helper functions.
+First of all one needs a function to handle the opening and combination of all parquet files.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def open_input_files
+:end-at: return all_events
+```
+
+In `prepare_events` we use the combined awkward and filter out all columns we are not interested in during the training, to stay lightweight.
```

Comment:
```suggestion
In `prepare_events` we use the combined awkward array and filter out all columns we are not interested in during the training, to stay lightweight.
```

Reply:
arguably, the most lightweight-way to load the columns would probably be in your `open_input_files` function

Reply:
Yes, it could be a good idea (from a didactic point of view) to move the filtering of the data to `open_input_files`, let `prepare_events` only handle thewhole preprocessing, if there is some.